packages/react/popover/src/Popover.tsx

```diff
@@ -167,15 +170,83 @@ const PopoverContent = React.forwardRef((props, forwardedRef) => {
 
   return (
     <Presence present={forceMount || context.open}>
-      <PopoverContentImpl
-        data-state={context.open ? 'open' : 'closed'}
-        {...contentProps}
-        ref={forwardedRef}
-      />
+      {context.modal ? (
+        <PopoverContentModal {...contentProps} ref={forwardedRef} />
+      ) : (
+        <PopoverContentNonModal {...contentProps} ref={forwardedRef} />
+      )}
     </Presence>
   );
 }) as PopoverContentPrimitive;
 
+PopoverContent.displayName = CONTENT_NAME;
+
+type PopoverContentTypeOwnProps = Omit<
+  Polymorphic.OwnProps<typeof PopoverContentImpl>,
+  'trapFocus' | 'disableOutsidePointerEvents'
+>;
+
+type PopoverContentTypePrimitive = Polymorphic.ForwardRefComponent<
+  Polymorphic.IntrinsicElement<typeof PopoverContentImpl>,
+  PopoverContentTypeOwnProps
+>;
+
+const PopoverContentModal = React.forwardRef((props, forwardedRef) => {
+  const context = usePopoverContext(CONTENT_NAME);
+  const contentRef = React.useRef<HTMLDivElement>(null);
+  const composedRefs = useComposedRefs(forwardedRef, contentRef);
+
+  // aria-hide everything except the content (better supported equivalent to setting aria-modal)
+  React.useEffect(() => {
+    const content = contentRef.current;
+    if (content) return hideOthers(content);
+  }, []);
+
+  return (
+    <PopoverContentImpl
+      {...props}
+      ref={composedRefs}
+      // we make sure we're not trapping once it's been closed
+      // (closed !== unmounted when animating out)
+      trapFocus={context.open}
+      disableOutsidePointerEvents
+      // When focus is trapped, a `focusout` event may still happen.
+      // We make sure we don't trigger our `onDismiss` in such case.
+      onFocusOutside={composeEventHandlers(props.onFocusOutside, (event) => event.preventDefault())}
+    />
```

Comment:
Does modal mode need right click handling as well?

Reply:
Discussed with the team and the answer here is no, we prevent right click dismiss on primitives with an `Overlay` part only e.g. `Dialog`. In that example we consider it contiguous with the primitive itself.